// @zilliz/milvus2-sdk-node/dist/milvus/types/Common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompactionState = exports.SegmentState = exports.DslType = exports.IndexState = exports.DataTypeMap = exports.DataType = exports.MsgType = exports.IndexType = exports.MetricType = void 0;
var MetricType;
(function (MetricType) {
    MetricType[MetricType["GHOST"] = 0] = "GHOST";
    // L2 euclidean distance
    MetricType[MetricType["L2"] = 1] = "L2";
    // IP inner product
    MetricType[MetricType["IP"] = 2] = "IP";
    // HAMMING hamming distance
    MetricType[MetricType["HAMMING"] = 3] = "HAMMING";
    // JACCARD jaccard distance
    MetricType[MetricType["JACCARD"] = 4] = "JACCARD";
    // TANIMOTO tanimoto distance
    MetricType[MetricType["TANIMOTO"] = 5] = "TANIMOTO";
    // SUBSTRUCTURE substructure distance
    MetricType[MetricType["SUBSTRUCTURE"] = 6] = "SUBSTRUCTURE";
    // SUPERSTRUCTURE superstructure
    MetricType[MetricType["SUPERSTRUCTURE"] = 7] = "SUPERSTRUCTURE";
})(MetricType = exports.MetricType || (exports.MetricType = {}));
var IndexType;
(function (IndexType) {
    // INVALID invald index type
    IndexType[IndexType["INVALID"] = 0] = "INVALID";
    // FLAT flat
    IndexType[IndexType["FLAT"] = 1] = "FLAT";
    // IVFFLAT ivfflat
    IndexType[IndexType["IVFFLAT"] = 2] = "IVFFLAT";
    // IVFSQ8 ivfsq8
    IndexType[IndexType["IVFSQ8"] = 3] = "IVFSQ8";
    //RNSG rnsg
    IndexType[IndexType["RNSG"] = 4] = "RNSG";
    // IVFSQ8H ivfsq8h
    IndexType[IndexType["IVFSQ8H"] = 5] = "IVFSQ8H";
    // IVFPQ ivfpq
    IndexType[IndexType["IVFPQ"] = 6] = "IVFPQ";
    // SPTAGKDT sptagkdt
    IndexType[IndexType["SPTAGKDT"] = 7] = "SPTAGKDT";
    // SPTAGBKT sptagbkt
    IndexType[IndexType["SPTAGBKT"] = 8] = "SPTAGBKT";
    // HNSW hnsw
    IndexType[IndexType["HNSW"] = 11] = "HNSW";
    // ANNOY annoy
    IndexType[IndexType["ANNOY"] = 12] = "ANNOY";
})(IndexType = exports.IndexType || (exports.IndexType = {}));
var MsgType;
(function (MsgType) {
    MsgType[MsgType["Undefined"] = 0] = "Undefined";
    /* DEFINITION REQUESTS: COLLECTION */
    MsgType[MsgType["CreateCollection"] = 100] = "CreateCollection";
    MsgType[MsgType["DropCollection"] = 101] = "DropCollection";
    MsgType[MsgType["HasCollection"] = 102] = "HasCollection";
    MsgType[MsgType["DescribeCollection"] = 103] = "DescribeCollection";
    MsgType[MsgType["ShowCollections"] = 104] = "ShowCollections";
    MsgType[MsgType["GetSystemConfigs"] = 105] = "GetSystemConfigs";
    MsgType[MsgType["LoadCollection"] = 106] = "LoadCollection";
    MsgType[MsgType["ReleaseCollection"] = 107] = "ReleaseCollection";
    /* DEFINITION REQUESTS: PARTITION */
    MsgType[MsgType["CreatePartition"] = 200] = "CreatePartition";
    MsgType[MsgType["DropPartition"] = 201] = "DropPartition";
    MsgType[MsgType["HasPartition"] = 202] = "HasPartition";
    MsgType[MsgType["DescribePartition"] = 203] = "DescribePartition";
    MsgType[MsgType["ShowPartitions"] = 204] = "ShowPartitions";
    MsgType[MsgType["LoadPartitions"] = 205] = "LoadPartitions";
    MsgType[MsgType["ReleasePartitions"] = 206] = "ReleasePartitions";
    /* DEFINE REQUESTS: SEGMENT */
    MsgType[MsgType["ShowSegments"] = 250] = "ShowSegments";
    MsgType[MsgType["DescribeSegment"] = 251] = "DescribeSegment";
    /* DEFINITION REQUESTS: INDEX */
    MsgType[MsgType["CreateIndex"] = 300] = "CreateIndex";
    MsgType[MsgType["DescribeIndex"] = 301] = "DescribeIndex";
    MsgType[MsgType["DropIndex"] = 302] = "DropIndex";
    /* MANIPULATION REQUESTS */
    MsgType[MsgType["Insert"] = 400] = "Insert";
    MsgType[MsgType["Delete"] = 401] = "Delete";
    MsgType[MsgType["Flush"] = 402] = "Flush";
    /* QUERY */
    MsgType[MsgType["Search"] = 500] = "Search";
    MsgType[MsgType["SearchResult"] = 501] = "SearchResult";
    MsgType[MsgType["GetIndexState"] = 502] = "GetIndexState";
    MsgType[MsgType["GetIndexBuildProgress"] = 503] = "GetIndexBuildProgress";
    MsgType[MsgType["GetCollectionStatistics"] = 504] = "GetCollectionStatistics";
    MsgType[MsgType["GetPartitionStatistics"] = 505] = "GetPartitionStatistics";
    MsgType[MsgType["Retrieve"] = 506] = "Retrieve";
    MsgType[MsgType["RetrieveResult"] = 507] = "RetrieveResult";
    /* DATA SERVICE */
    MsgType[MsgType["SegmentInfo"] = 600] = "SegmentInfo";
    /* SYSTEM CONTROL */
    MsgType[MsgType["TimeTick"] = 1200] = "TimeTick";
    MsgType[MsgType["QueryNodeStats"] = 1201] = "QueryNodeStats";
    MsgType[MsgType["LoadIndex"] = 1202] = "LoadIndex";
    MsgType[MsgType["RequestID"] = 1203] = "RequestID";
    MsgType[MsgType["RequestTSO"] = 1204] = "RequestTSO";
    MsgType[MsgType["AllocateSegment"] = 1205] = "AllocateSegment";
    MsgType[MsgType["SegmentStatistics"] = 1206] = "SegmentStatistics";
    MsgType[MsgType["SegmentFlushDone"] = 1207] = "SegmentFlushDone";
})(MsgType = exports.MsgType || (exports.MsgType = {}));
/**
 * @brief Field data type
 */
var DataType;
(function (DataType) {
    DataType[DataType["None"] = 0] = "None";
    DataType[DataType["Bool"] = 1] = "Bool";
    DataType[DataType["Int8"] = 2] = "Int8";
    DataType[DataType["Int16"] = 3] = "Int16";
    DataType[DataType["Int32"] = 4] = "Int32";
    DataType[DataType["Int64"] = 5] = "Int64";
    DataType[DataType["Float"] = 10] = "Float";
    DataType[DataType["Double"] = 11] = "Double";
    DataType[DataType["String"] = 20] = "String";
    DataType[DataType["BinaryVector"] = 100] = "BinaryVector";
    DataType[DataType["FloatVector"] = 101] = "FloatVector";
})(DataType = exports.DataType || (exports.DataType = {}));
exports.DataTypeMap = {
    none: 0,
    bool: 1,
    int8: 2,
    int16: 3,
    int32: 4,
    int64: 5,
    float: 10,
    double: 11,
    string: 20,
    binaryvector: 100,
    floatvector: 101,
};
var IndexState;
(function (IndexState) {
    IndexState[IndexState["IndexStateNone"] = 0] = "IndexStateNone";
    IndexState[IndexState["Unissued"] = 1] = "Unissued";
    IndexState[IndexState["InProgress"] = 2] = "InProgress";
    IndexState[IndexState["Finished"] = 3] = "Finished";
    IndexState[IndexState["Failed"] = 4] = "Failed";
})(IndexState = exports.IndexState || (exports.IndexState = {}));
var DslType;
(function (DslType) {
    DslType[DslType["Dsl"] = 0] = "Dsl";
    DslType[DslType["BoolExprV1"] = 1] = "BoolExprV1";
})(DslType = exports.DslType || (exports.DslType = {}));
var SegmentState;
(function (SegmentState) {
    SegmentState[SegmentState["SegmentStateNone"] = 0] = "SegmentStateNone";
    SegmentState[SegmentState["NotExist"] = 1] = "NotExist";
    SegmentState[SegmentState["Growing"] = 2] = "Growing";
    SegmentState[SegmentState["Sealed"] = 3] = "Sealed";
    SegmentState["Flushed"] = "Flushed";
    SegmentState["Flushing"] = "Flushing";
})(SegmentState = exports.SegmentState || (exports.SegmentState = {}));
var CompactionState;
(function (CompactionState) {
    CompactionState[CompactionState["UndefiedState"] = 0] = "UndefiedState";
    CompactionState[CompactionState["Executing"] = 1] = "Executing";
    CompactionState[CompactionState["Completed"] = 2] = "Completed";
})(CompactionState = exports.CompactionState || (exports.CompactionState = {}));
